python/featomic/tests/clebsch_gordan/cartesian_spherical.py

```diff
@@ -198,3 +200,256 @@ def test_cartesian_to_spherical_errors(cartesian):
+    )
+
+    # Extract the lambda = 2 component
+    l2_input = operations.drop_blocks(
+        operations.remove_dimension(rank_2_input_sph, "keys", "_"),
+        keys=Labels(["o3_lambda"], np.array([[0], [1]])),
+    )
+
+    assert operations.equal_metadata(l2_input, l2_reference)
+    assert operations.allclose(l2_input, l2_reference)
+
+
+@pytest.mark.parametrize("cg_backend", ["python-dense", "python-sparse"])
+def test_cartesian_to_spherical_rank_3_by_equation(cg_backend):
+    """
+    Tests cartesian_to_spherical for a random rank-2 tensor by comparing the result to
+    the result from *the* reference equation.
+    """
+    # Build the reference lambda = 2 component
+    random_rank_3_arr = np.random.rand(100, 3, 3, 3, 1)
+    l3_reference = TensorMap(
+        keys=Labels(["o3_lambda", "o3_sigma"], np.array([[3, 1]])),
+        blocks=[
+            TensorBlock(
+                samples=Labels(["system"], np.arange(100).reshape(-1, 1)),
+                components=[Labels(["o3_mu"], np.arange(-3, 4).reshape(-1, 1))],
+                properties=Labels(["_"], np.array([[0]])),
+                values=np.stack(
+                    [_l3_components_from_matrix(A[..., 0]) for A in random_rank_3_arr]
+                ).reshape(random_rank_3_arr.shape[0], 7, 1),
+            )
+        ],
+    )
+
+    # Build the cartesian tensor and do cartesian to spherical
+    rank_3_input_cart = TensorMap(
+        keys=Labels.single(),
+        blocks=[
+            TensorBlock(
+                values=random_rank_3_arr,
+                samples=Labels(["system"], np.arange(100).reshape(-1, 1)),
+                components=[
+                    Labels([a], np.arange(3).reshape(-1, 1))
+                    for a in ["xyz1", "xyz2", "xyz3"]
+                ],
+                properties=Labels(["_"], np.array([[0]])),
+            )
+        ],
+    )
+    rank_3_input_sph = cartesian_to_spherical(
+        rank_3_input_cart, ["xyz1", "xyz2", "xyz3"], cg_backend=cg_backend
+    )
+
+    # Extract the lambda = 3 component
+    l3_input = operations.drop_blocks(
+        operations.remove_dimension(rank_3_input_sph, "keys", "_"),
+        keys=Labels(["o3_lambda"], np.array([[0], [1], [2]])),
+    )
+    for dim in ["l_3", "k_1", "l_2", "l_1"]:
+        l3_input = operations.remove_dimension(l3_input, "keys", dim)
+
+    assert operations.equal_metadata(l3_input, l3_reference)
+    assert operations.allclose(l3_input, l3_reference)
+
+
+@pytest.mark.parametrize("cg_backend", ["python-dense", "python-sparse"])
+def test_cartesian_to_spherical_equivariance(cg_backend):
+    # Define a random cartesian rank 2 tensor
+    random_rank_2_arr = np.random.rand(100, 3, 3, 1)
+
+    # Define some rotation angles, the cartesian rotation matrix, and the Wigner
+    # matrices
+    angles = np.random.randn(3) * np.pi
+    R = cartesian_rotation(angles)
+    wigner = WignerDReal(max_angular=2, angles=angles)
```

Comment:
I am now a bit afraid that the code in WignerDReal might also be wrong. Was it also validated on the side?

Reply:
Hmm good point. @MichelangeloDomina please could you work your magic?